Library/Homebrew/service.rb

```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
 
       timer + options.join("\n")
     end
+
+    # Only evaluate the service block once.
+    sig { void }
+    def eval_service_block
+      return if @eval_service_block
+
+      instance_eval(&@service_block)
+      @eval_service_block = true
+    end
+
+    # Recursively prepare the service hash for inclusion in the formula API JSON.
+    # - Replace each Symbol with a String prefixed by ":"
+    # - Replace the local HOMEBREW_PREFIX with "$HOMEBREW_PREFIX"
+    # - Replace the local home directory with "$HOME"
+    def serialize(elem = to_h)
+      case elem
+      when String, Pathname
+        elem.to_s.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
+            .gsub(Dir.home, "$HOME")
+      when Symbol
+        elem.inspect
+      when Array
+        elem.map { |value| serialize(value) }
+      when Hash
+        elem.to_h do |key, value|
+          key = key.inspect if key.is_a?(Symbol)
+          [key, serialize(value)]
+        end
+      else
+        elem
+      end
+    end
+
+    # Recursively turn the service API hash values back into what is expected by the formula DSL.
+    # - Replace each String prefixed with ":" with a Symbol
+    # - Replace "$HOMEBREW_PREFIX" with the local HOMEBREW_PREFIX environment variable
+    # - Replace "$HOME" with the local home directory
+    def self.deserialize(elem)
```

Comment:
`from_hash` feels more typical naming here

Reply:
I guess I'd expect `from_hash` to create a service object which is not what's happening here. I'm not really happy with the naming here either though.